### ISSUE TITLE
The error was: ModuleNotFoundError: No module named 'dateutil'

### ISSUE THREAD

**Issue by smatterchew** [ashleyghooper/solarwinds_ansible#1]
Ansible version:
```
ansible [core 2.11.12] 
  config file = /home/user/.ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/venv/ansible/lib64/python3.6/site-packages/ansible
  ansible collection location = /home/user/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/user/venv/ansible/bin/ansible
  python version = 3.6.8 (default, Aug 13 2020, 07:46:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
  jinja version = 3.0.3
  libyaml = True
```
Collection version:
```
# /home/user/.ansible/collections/ansible_collections
Collection                  Version
--------------------------- -------
anophelesgreyhoe.solarwinds 1.2.0  
```
Python modules installed:
```
ansible-core==2.11.12
ansible-lint==5.4.0
bracex==2.2.1
certifi==2022.9.24
cffi==1.15.1
charset-normalizer==2.0.12
commonmark==0.9.1
cryptography==38.0.1
dataclasses==0.8
enrich==1.2.7
idna==3.4
Jinja2==3.0.3
jmespath==0.10.0
MarkupSafe==2.0.1
orionsdk==0.3.0
packaging==21.3
pathspec==0.9.0
pycparser==2.21
Pygments==2.13.0
pyparsing==3.0.9
python-dateutil==2.8.2
PyYAML==6.0
requests==2.27.1
resolvelib==0.5.4
rich==12.5.1
ruamel.yaml==0.17.21
ruamel.yaml.clib==0.2.6
six==1.16.0
tenacity==8.1.0
typing_extensions==4.1.1
urllib3==1.26.12
wcmatch==8.3
yamllint==1.28.0
```
Attempt to install the `dateutil` module:
```
[user@localhost] ➤ pip install dateutil
ERROR: Could not find a version that satisfies the requirement dateutil (from versions: none)
ERROR: No matching distribution found for dateutil
```

Looking around for a solution to this error, I found [this link](
https://techoverflow.net/2019/10/21/how-to-fix-modulenotfounderror-no-module-named-dateutil/) to install `python-dateutil` but it isn't helping.

What else can I provide to help fix this?

Thank you

**Comment by ashleyghooper** [comment]
Hi there,
when you say the link to install `python-dateutil` isn't helping, do you mean that you were able to install it okay but that it didn't help?

I attempted to replicate your issue in a docker container (`ubuntu:22.04`) from both Python and Ansible. In both cases, installing the `dateutil` module fixed the issue.

In my Ubuntu container, I was able to install the dependency by either of the following methods:

```sh
pip install python-dateutil
```

...or...

```sh
apt install python3-dateutil
```

Please let me know if you have installed the `python-dateutil` module such that it's available to the user running Ansible, and are still seeing the missing dependency.

# Output

## Python

```sh
root@10b3d1ede0ae:~/test# python3 -c 'import dateutil'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'dateutil'
root@10b3d1ede0ae:~/test# pip install python-dateutil
Collecting python-dateutil
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil) (1.16.0)
Installing collected packages: python-dateutil
Successfully installed python-dateutil-2.8.2
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
root@10b3d1ede0ae:~/test# python3 -c 'import dateutil'
root@10b3d1ede0ae:~/test#
```

## Ansible

Then I replicated from Ansible with a bogus playbook as below.

```yaml
- name: Add node to SolarWinds
  hosts: all
  gather_facts: no
  tasks:

    - name: Discover and add node to Orion if not already
      anophelesgreyhoe.solarwinds.orion_node:
        solarwinds_connection:
          hostname: nonexistent.com
          username: username
          password: password
        node_name: node
        caption: caption
        state: present
        ip_address: 192.168.100.100
        polling_method: snmp
        polling_engine_name: polling.com
        discovery_polling_engine_name:
        snmp_port: 161
        snmp_version: '3'
        snmp_allow_64: yes
        credential_name: a_cred
        interface_filters:
        volume_filters:
        volume_filter_cutoff_max: 10
        custom_properties:
      delegate_to: localhost
      throttle: 1
```

```sh
root@10b3d1ede0ae:~/test# ansible-playbook -i localhost, main.yml

PLAY [Add node to SolarWinds] ***********************************************************

TASK [Discover and add node to Orion if not already] ************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ModuleNotFoundError: No module named 'dateutil'
fatal: [localhost -> localhost]: FAILED! => {"changed": false, "msg": "Failed to import the required Python library (dateutil) on 10b3d1ede0ae's Python /usr/bin/python3. Please read the module documentation and install it in the appropriate location. If the required library is installed, but Ansible is using the wrong Python interpreter, please consult the documentation on ansible_python_interpreter"}

PLAY RECAP ******************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

root@10b3d1ede0ae:~/test# pip install python-dateutil
Collecting python-dateutil
  Downloading python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 247.7/247.7 KB 1.1 MB/s eta 0:00:00
Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil) (1.16.0)
Installing collected packages: python-dateutil
Successfully installed python-dateutil-2.8.2
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
root@10b3d1ede0ae:~/test# ansible-playbook -i localhost, main.yml

PLAY [Add node to SolarWinds] ***********************************************************

TASK [Discover and add node to Orion if not already] ************************************
fatal: [localhost -> localhost]: FAILED! => {"changed": false, "msg": "Failed to query Orion. Check Orion hostname, ********, and/or ********: HTTPSConnectionPool(host='nonexistent.com', port=17778): Max retries exceeded with url: /SolarWinds/InformationService/v3/Json/Query (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa3653b8550>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))"}

PLAY RECAP ******************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by ashleyghooper** [comment]
@smatterchew, I see from your list of Python modules installed that you already have `python-dateutil==2.8.2`, which is the same version I have.

Can you run `ansible --version` to confirm which Python interpreter is being used by Ansible. We then need to make sure that the `dateutil` module is available to that Python interpreter for the user running Ansible.

**Comment by smatterchew** [comment]
Hi @ashleyghooper.  The ansible version is in the original message above.

**Comment by ashleyghooper** [comment]
> Hi @ashleyghooper. The ansible version is in the original message above.

Sorry, right you are! It was the _python_ interpreter version I was after, not Ansible, but yes, it is there in the output you had already included, so we can see there that it's Python 3.6.8.

```
  python version = 3.6.8 (default, Aug 13 2020, 07:46:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

Assuming your python 3 interpreter is in your path and is named `python3`, you should be able to verify that the `dateutil` module is available to it like so from a bash shell (assuming bash is new enough to support HEREDOCs):

```
python3 <<HEREDOC
try:
    import dateutil
    print("success!")
except Exception:
    print("import failed")
HEREDOC
```

**Comment by smatterchew** [comment]
Hi @ashleyghooper.  Your question pointed me in the right direction here.  I got a `success` from your test.

The solution was to create an inventory file that pointed to the venv I am using:
```
cat ~/inventories/localhost.ansible
localhost ansible_connection=local ansible_python_interpreter=~/venv/ansible/bin/python
```

I think I may still be missing something, possibly in my venv, because I'm getting a traceback that eventually points to line 362 in `orion_node.py`:
```
AttributeError: 'NoneType' object has no attribute 'upper'
```

The line in `orion_node.py` reads: 
```python
    'ObjectSubType': params['polling_method'].upper(),
```

Should I close this issue and open another?

**Comment by smatterchew** [comment]
A little more tinkering.

Appears that I don't have the `polling_method` set in my call to the module and `polling_method` is not a required parm.

```yaml
    - name: "Add a record to Orion"
      delegate_to: localhost
      anophelesgreyhoe.solarwinds.orion_node:
        solarwinds_connection:
          hostname: "{{ survey_solarwinds_host }}"
          username: "{{ survey_username }}"
          password: "{{ survey_password }}"
        state: present
        node_name: "labs99999.contoso.com"
        caption: "This is the caption"
        snmp_version: 3
```
When I set the `polling_method: snmp` the error goes away.

**Comment by ashleyghooper** [comment]
Hi @smatterchew, yes `polling_method` is only required when adding a new node, i.e. `state: present`, but not when performing any of the other operations, i.e. `mute`, `unmute`, `unmanage`, `remanage`, etc.

That might be an indication that we could have better scoping of this module, and split off the mute, unmute, etc into another module, although not sure it'd be worth it.

For now we can do better in terms of error handling though. I'll add explicit checks for all of the required parameters when `state=present`.

Did you get the node added to SolarWinds in the end? I think you would also have needed to indicate an existing SNMP3 credential you have in SolarWinds that matched the `snmpd` config on the node you're adding via the `credential_name` parameter.

**Comment by smatterchew** [comment]
Hi @ashleyghooper.  Yes, I am able to get the node added and removed now. Thank you for the work you put into this collection.  The `orion_node` module will allow us to bypass one more team in the provisioning/decommissioning processes.  Very nice.

**Comment by ashleyghooper** [comment]
> Hi @ashleyghooper. Yes, I am able to get the node added and removed now.

That's great to hear. Thanks for the feedback.

> Thank you for the work you put into this collection. The `orion_node` module will allow us to bypass one more team in the provisioning/decommissioning processes. Very nice.

I'm really glad to hear it's helping somebody. And I have to give a lot of credit to @jdchaiken on whose [work](https://github.com/jdchaiken/solarwinds-Node_Module) this is based.

**Comment by ashleyghooper** [comment]
Speaking of the original project, when I added the SolarWinds network discovery code, I wasn't aware of the possibility of calling `ScheduleListResources` followed by `ImportListResourcesResult` to inventory the machine and automatically add pollers for everything. A lot of googling and perusal of the SolarWinds forums had led me to believe that network discovery was the only way to automatically have pollers created.

In my recent work adding passive agent support, I discovered the above alternative, so I was intending to try reverting back to simply adding a node using something like the original method from @jdchaiken module, then using the list/import approach.

If that works, I'd probably split discovery into a separate module that could be used for actual discovery of a whole network range as intended, whereas right now it's being used to 'discover' a single IP.
